File: gen_files.js
#!/usr/bin/env node
/**
 * 扫描目录下所有研报 .md/.html 文件，自动生成 files.json 清单
 * 用法：node gen_files.js
 */

const fs = require('fs');
const path = require('path');

const ROOT = __dirname;

// 不纳入清单的文件（配置类）
const SKIP = new Set([
  'README.md', 'CLOUDFLARE_SETUP.md', 'DEPLOY.md', 'index.html',
  '_headers', '.gitignore', 'gen_files.js', 'gen_files.py', 'files.json'
]);

// 关键词 → 标签映射
const TAG_MAP = [
  { keywords: ['戴尔', 'DELL', 'dell'], tag: '戴尔' },
  { keywords: ['Anthropic', 'anthropic', 'ANTHROPIC'], tag: 'Anthropic' },
  { keywords: ['卖方', '券商', '晨会'], tag: '券商晨会' },
  { keywords: ['特朗普', 'trump', 'Trump'], tag: '特朗普' },
  { keywords: ['AI', '算力', '服务器', 'GPU', 'TPU', '光模块', 'PCB'], tag: 'AI算力' },
  { keywords: ['持仓', '盈亏'], tag: '持仓分析' },
  { keywords: ['概念', '受益', '产业链'], tag: '概念股' },
  { keywords: ['个股', '梳理'], tag: '个股梳理' },
];

function extractTitle(filepath, content) {
  // MD: 找第一个 # 标题
  let m = content.match(/^#\s+(.+)/m);
  if (m) return m[1].trim();
  // HTML: <title> 或 <h1>
  m = content.match(/<title>(.+?)<\/title>/i);
  if (m) return m[1].trim();
  m = content.match(/<h1[^>]*>(.+?)<\/h1>/i);
  if (m) return m[1].trim();
  return path.basename(filepath, path.extname(filepath));
}

function extractDate(filepath) {
  const filename = path.basename(filepath);
  // 标准日期 YYYY-MM-DD
  let m = filename.match(/(\d{4}-\d{2}-\d{2})/);
  if (m) return m[1];
  // 兜底：文件修改时间
  const stat = fs.statSync(filepath);
  const d = new Date(stat.mtime);
  return d.getFullYear() + '-' +
    String(d.getMonth() + 1).padStart(2, '0') + '-' +
    String(d.getDate()).padStart(2, '0');
}

function extractDesc(filepath, content) {
  const ext = path.extname(filepath).toLowerCase();
  try {
    if (ext === '.md') {
      const lines = content.split('\n');
      for (const line of lines) {
        const s = line.trim();
        if (!s || s.startsWith('#') || s.startsWith('>') || s.startsWith('---')) continue;
        if (s.startsWith('|')) continue;
        if (s.length > 20) return cutText(s, 100);
        if (s.length > 10) return cutText(s, 100);
      }
    } else if (ext === '.html') {
      let m = content.match(/<meta\s+name="description"\s+content="(.+?)"/i);
      if (m) return m[1].slice(0, 100);
      m = content.match(/<p[^>]*>(.+?)<\/p>/i);
      if (m) {
        const text = m[1].replace(/<[^>]+>/g, '').trim();
        if (text.length > 20) return cutText(text, 100);
      }
    }
  } catch (e) { /* ignore */ }
  return '';
}

function cutText(text, maxLen) {
  text = text.replace(/\*\*/g, '');
  if (text.length <= maxLen) return text;
  return text.slice(0, maxLen).replace(/[，,。、；;：:！!？?]+$/, '') + '…';
}

function autoTags(title, desc, filename) {
  const combined = (title + ' ' + desc + ' ' + filename).toLowerCase();
  const tags = [];
  for (const { keywords, tag } of TAG_MAP) {
    for (const kw of keywords) {
      if (combined.includes(kw.toLowerCase())) {
        if (!tags.includes(tag)) tags.push(tag);
        break;
      }
    }
  }
  return tags;
}

function scan() {
  const items = [];
  const files = fs.readdirSync(ROOT);

  for (const name of files.sort()) {
    if (SKIP.has(name)) continue;
    const ext = path.extname(name).toLowerCase();
    if (ext !== '.md' && ext !== '.html') continue;

    const filepath = path.join(ROOT, name);
    const content = fs.readFileSync(filepath, 'utf-8');
    const title = extractTitle(filepath, content);
    const date = extractDate(filepath);
    const desc = extractDesc(filepath, content);
    const tags = autoTags(title, desc, name);
    const ftype = ext.slice(1); // 'md' or 'html'

    items.push({
      filename: name,
      title,
      date,
      desc: desc || title,
      tags,
      type: ftype,
      href: name
    });
  }

  return items;
}

function merge(items) {
  const merged = {};
  for (const item of items) {
    const base = item.filename.replace(/\.(md|html)$/i, '');
    if (!merged[base]) {
      merged[base] = {
        title: item.title,
        date: item.date,
        desc: item.desc,
        tags: item.tags,
        files: []
      };
    }
    const m = merged[base];
    merged[base].files.push({
      name: item.type.toUpperCase(),
      href: item.href,
      type: item.type
    });
    // 取更优的字段
    if (item.title.length > m.title.length) m.title = item.title;
    if (item.desc.length > m.desc.length) m.desc = item.desc;
    if (item.tags.length > m.tags.length) m.tags = item.tags;
    if (item.date && !m.date) m.date = item.date;
  }
  const result = Object.values(merged);
  result.sort((a, b) => b.date.localeCompare(a.date));
  return result;
}

function main() {
  const items = scan();
  const result = merge(items);

  const outputPath = path.join(ROOT, 'files.json');
  fs.writeFileSync(outputPath, JSON.stringify(result, null, 2), 'utf-8');

  console.log(`[OK] 扫描完成，共 ${result.length} 个文档条目 → ${outputPath}`);
  for (const item of result) {
    const ftypes = item.files.map(f => f.name).join('/');
    console.log(`  📄 ${item.title}  [${ftypes}]  ${item.date}`);
  }
}

main();

File: gen_files.py
#!/usr/bin/env python3
"""扫描目录下所有研报 .md/.html 文件，自动生成 files.json 清单"""

import re
import json
import os
from pathlib import Path
from datetime import datetime

ROOT = Path(__file__).parent

# 不纳入清单的文件（配置类）
SKIP = {
    "README.md", "CLOUDFLARE_SETUP.md", "index.html", "_headers",
    ".gitignore", "gen_files.py"
}

# 关键词 → 标签映射
TAG_MAP = [
    (["戴尔", "DELL", "dell"], "戴尔"),
    (["Anthropic", "anthropic", "anthropic"], "Anthropic"),
    (["卖方", "券商", "晨会"], "券商晨会"),
    (["特朗普", "trump", "Trump"], "特朗普"),
    (["AI", "算力", "服务器", "GPU", "TPU", "光模块", "PCB"], "AI算力"),
    (["持仓", "盈亏"], "持仓分析"),
    (["概念", "受益", "产业链"], "概念股"),
    (["个股", "梳理"], "个股梳理"),
]

def extract_title(filepath):
    """从 MD/HTML 文件中提取标题"""
    try:
        content = filepath.read_text(encoding="utf-8")
        # MD: 找第一个 # 标题
        m = re.search(r"^#\s+(.+)", content, re.MULTILINE)
        if m:
            return m.group(1).strip()
        # HTML: 找 <title> 或第一个 <h1>
        m = re.search(r"<title>(.+?)</title>", content, re.IGNORECASE)
        if m:
            return m.group(1).strip()
        m = re.search(r"<h1[^>]*>(.+?)</h1>", content, re.IGNORECASE)
        if m:
            return m.group(1).strip()
    except Exception:
        pass
    return filepath.stem

def extract_date(filepath):
    """从文件名中提取日期 (YYYY-MM-DD)，失败则用文件修改时间"""
    filename = filepath.name
    # 优先匹配标准日期格式
    m = re.search(r"(\d{4}-\d{2}-\d{2})", filename)
    if m:
        return m.group(1)
    # 兜底：文件修改时间
    ts = filepath.stat().st_mtime
    return datetime.fromtimestamp(ts).strftime("%Y-%m-%d")

def extract_desc(filepath):
    """从文件中提取简短描述（前 100 个有效字符）"""
    try:
        content = filepath.read_text(encoding="utf-8")
        suffix = filepath.suffix.lower()
        if suffix == ".md":
            # 跳过标题行和空行，取第一个非空段落
            lines = content.split("\n")
            in_text = False
            for line in lines:
                stripped = line.strip()
                if not stripped or stripped.startswith("#") or stripped.startswith(">"):
                    continue
                if stripped.startswith("---"):
                    continue
                if len(stripped) > 20 and not stripped.startswith("|"):
                    return stripped[:100].rstrip("。，、；：！？") + "…"
                if len(stripped) > 10:
                    return stripped[:100].rstrip("。，、；：！？") + "…"
        elif suffix == ".html":
            # 找 meta description 或第一个有内容的 p
            m = re.search(r'<meta\s+name="description"\s+content="(.+?)"', content, re.IGNORECASE)
            if m:
                return m.group(1)[:100]
            m = re.search(r"<p[^>]*>(.+?)</p>", content, re.IGNORECASE)
            if m:
                text = re.sub(r"<[^>]+>", "", m.group(1)).strip()
                if len(text) > 20:
                    return text[:100].rstrip("。，、；：！？") + "…"
    except Exception:
        pass
    return ""

def auto_tags(title, desc, filename):
    """根据标题、描述、文件名自动生成标签"""
    combined = f"{title} {desc} {filename}".lower()
    tags = []
    for keywords, tag in TAG_MAP:
        for kw in keywords:
            if kw.lower() in combined:
                if tag not in tags:
                    tags.append(tag)
                break
    return tags

def scan():
    items = []
    for f in sorted(ROOT.iterdir()):
        if not f.is_file():
            continue
        if f.name in SKIP:
            continue
        suffix = f.suffix.lower()
        if suffix not in (".md", ".html"):
            continue

        title = extract_title(f)
        date = extract_date(f)
        desc = extract_desc(f)
        tags = auto_tags(title, desc, f.name)
        ftype = suffix.lstrip(".")

        items.append({
            "filename": f.name,
            "title": title,
            "date": date,
            "desc": desc or f"{title}",
            "tags": tags,
            "type": ftype,
            "href": f.name
        })

    return items

def main():
    items = scan()

    # 去重：同一个主题的 md+html 合并为一个条目
    merged = {}
    for item in items:
        # 用文件名去掉扩展名的部分作为 key
        base = re.sub(r"\.(md|html)$", "", item["filename"], flags=re.IGNORECASE)
        if base not in merged:
            merged[base] = {
                "title": item["title"],
                "date": item["date"],
                "desc": item["desc"],
                "tags": item["tags"],
                "files": []
            }
        merged[base]["files"].append({
            "name": item["type"].upper(),
            "href": item["href"],
            "type": item["type"]
        })
        # 取更长的标题/描述
        if len(item["title"]) > len(merged[base]["title"]):
            merged[base]["title"] = item["title"]
        if len(item["desc"]) > len(merged[base]["desc"]):
            merged[base]["desc"] = item["desc"]
        if len(item["tags"]) > len(merged[base]["tags"]):
            merged[base]["tags"] = item["tags"]
        if item["date"] and not merged[base]["date"]:
            merged[base]["date"] = item["date"]

    result = list(merged.values())

    # 按日期倒序
    result.sort(key=lambda x: x["date"], reverse=True)

    output_path = ROOT / "files.json"
    output_path.write_text(
        json.dumps(result, ensure_ascii=False, indent=2),
        encoding="utf-8"
    )

    print(f"[OK] 扫描完成，共 {len(result)} 个文档条目 → {output_path}")
    for item in result:
        ftypes = "/".join(f["name"] for f in item["files"])
        print(f"  📄 {item['title']}  [{ftypes}]  {item['date']}")

if __name__ == "__main__":
    main()
